caddyhttp/proxy/reverseproxy.go

```diff
@@ -154,7 +155,12 @@ func NewSingleHostReverseProxy(target *url.URL, without string, keepalive int) *
 				prefer(target.RawPath, target.Path),
 				prefer(req.URL.RawPath, req.URL.Path))
 		}
-		req.URL.Path = singleJoiningSlash(target.Path, req.URL.Path)
+		req.URL.Path = singleJoiningSlash(target.Path,
+			prefer(req.Context().Value(staticfiles.URLPathCtxKey).(string), req.URL.Path))
+		// req.URL.Path must be consistent with decoded form of req.URL.RawPath if any
+		if req.URL.RawPath != "" && req.URL.RawPath != req.URL.EscapedPath() {
+			panic("RawPath doesn't match Path")
```

Comment:
@mholt I still think we should panic here because if the final request's `RawPath` and `Path` are not consistent. There must be a bug in our director logic and we should recognized it ASAP.
BTW I have add a test in the new patchset.